taropsbt/decode.go

```diff
@@ -0,0 +1,377 @@
+	"github.com/lightninglabs/taro/address"
+	"github.com/lightninglabs/taro/asset"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+// NewFromRawBytes returns a new instance of a VPacket struct created by reading
+// from a byte slice. If the format is invalid, an error is returned. If the
+// argument b64 is true, the passed byte slice is decoded from base64 encoding
+// before processing.
+func NewFromRawBytes(r io.Reader, b64 bool) (*VPacket, error) {
+	packet, err := psbt.NewFromRawBytes(r, b64)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding PSBT: %w", err)
+	}
+
+	return NewFromPsbt(packet)
+}
+
+// NewFromPsbt returns a new instance of a VPacket struct created by reading the
+// custom fields on the given PSBT packet.
+func NewFromPsbt(packet *psbt.Packet) (*VPacket, error) {
+	// Make sure we have the correct markers for a virtual transaction.
+	if len(packet.Unknowns) != 2 {
+		return nil, fmt.Errorf("expected 2 global unknown fields, "+
+			"got %d", len(packet.Unknowns))
+	}
+
+	// We want an explicit "isVirtual" boolean marker.
+	isVirtual, err := value(
+		packet.Unknowns, PsbtKeyTypeGlobalTaroIsVirtualTx,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error checking if virtual tx: %w", err)
+	}
+	if !bytes.Equal(isVirtual, trueAsBytes) {
+		return nil, fmt.Errorf("not a virtual transaction")
+	}
+
+	// We also want the HRP of the Taro chain params.
+	hrp, err := value(packet.Unknowns, PsbtKeyTypeGlobalTaroChainParamsHRP)
+	if err != nil {
+		return nil, fmt.Errorf("error reading Taro chain params HRP: "+
+			"%w", err)
+	}
+	chainParams, err := address.Net(string(hrp))
+	if err != nil {
+		return nil, fmt.Errorf("error parsing Taro chain params HRP: "+
+			"%w", err)
+	}
+
+	// Currently only one input is supported.
+	if len(packet.Inputs) != 1 {
+		return nil, fmt.Errorf("expected exactly one input, got %d",
+			len(packet.Inputs))
+	}
+	if len(packet.UnsignedTx.TxIn) != 1 {
+		return nil, fmt.Errorf("expected exactly one unsigned input, "+
+			"got %d", len(packet.UnsignedTx.TxIn))
+	}
+
+	vPacket := &VPacket{
+		ChainParams: chainParams,
+		Input:       &VInput{},
+		Outputs:     make([]*VOutput, len(packet.Outputs)),
+	}
+
+	err = vPacket.Input.decode(packet.Inputs[0])
+	if err != nil {
+		return nil, fmt.Errorf("error decoding virtual input: %w", err)
+	}
+
+	for idx := range packet.Outputs {
+		vOut := &VOutput{}
+		err := vOut.decode(
+			packet.Outputs[idx], packet.UnsignedTx.TxOut[idx],
+		)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding virtual output "+
+				"%d: %w", idx, err)
+		}
+
+		vPacket.Outputs[idx] = vOut
+	}
+
+	return vPacket, nil
+}
+
+// decode decodes the given PInput into the current VInput.
+func (i *VInput) decode(pIn psbt.PInput) error {
+	i.PInput = pIn
+
+	var (
+		prevID            *asset.PrevID
+		anchorValue       uint64
+		anchorSigHashType uint64
+	)
+
+	mapping := []struct {
+		key     []byte
+		decoder func([]byte) error
+	}{{
+		key:     PsbtKeyTypeInputTaroPrevID,
```

Comment:
TLV within PSBT 👽

Def a fan of, as the OG PSBT serialization format is a bit wonky.